File: 2096. Step-By-Step Directions From a Binary Tree Node to Another.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public String getDirections(TreeNode root, int startValue, int destValue) {
        StringBuilder source = new StringBuilder();
        StringBuilder target = new StringBuilder();
        getDirectionsFromRoot(root, startValue, source);
        getDirectionsFromRoot(root, destValue, target);
        source.reverse();
        target.reverse();
        int i = 0;
        while (i < source.length() && i < target.length()) {
            if (source.charAt(i) == target.charAt(i)) {
                i++;
            } else {
                break;
            }
        }
        StringBuilder result = new StringBuilder("U".repeat(source.length()-i));
        if (i != target.length()) {
            result.append(target.substring(i, target.length()));
        }
        return result.toString();
    }


    private boolean getDirectionsFromRoot(TreeNode root, int target, StringBuilder current) {
        if (root.val == target) {
            return true;
        }
        if (root.left != null && getDirectionsFromRoot(root.left, target, current)) {
            current.append("L");
        } else if (root.right != null && getDirectionsFromRoot(root.right, target, current)) {
            current.append("R");
        }
        return current.length() > 0;
    }
}

File: 2096. Step-By-Step Directions From a Binary Tree Node to Another.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func getDirections(root *TreeNode, startValue int, destValue int) string {
    path1, _ := find(root, startValue, []byte{})
    path2, _ := find(root, destValue, []byte{})
    for len(path1) > 0 && len(path2) > 0 && path1[0] == path2[0] {
        path1, path2 = path1[1:], path2[1:]
    }
    return strings.Repeat("U", len(path1)) + string(path2)
}

func find(root *TreeNode, val int, s []byte) ([]byte, bool) {
    if root == nil {
        return nil, false
    }
    if root.Val == val {
        return s, true
    }
    temp := append(s, 'L')
    if result, found := find(root.Left, val, temp); found {
        return result, true
    }
    return find(root.Right, val, append(s, 'R'))
}
